actionpack/lib/action_dispatch/http/url.rb

```diff
@@ -202,15 +202,15 @@ def extract_subdomains_from(host, tld_length)
 
           def build_host_url(host, port, protocol, options, path)
             if match = host.match(HOST_REGEXP)
-              protocol ||= match[1] unless protocol == false
-              host       = match[2]
-              port       = match[3] unless options.key? :port
+              protocol_from_host = match[1] if protocol.nil?
+              host               = match[2]
+              port               = match[3] unless options.key? :port
             end
 
-            protocol = normalize_protocol protocol
+            protocol = protocol_from_host || normalize_protocol(protocol)
```

Comment:
Previously this would always normalize `protocol`
Does this skip normalizing `protocol_from_host` now when the regexp matches?